### PR TITLE
buildkite: Generate junit xml file for logged errors

### PR DESCRIPTION
**Pull request by def-**
The motivation is to reduce manual effort to look into logs to figure
out that an issue is still occurring. The junit-annotate buildkite
plugin will use this information to annotate the change. The junit xml
is also uploaded to test analytics for future analysis.

Fixes: https://github.com/MaterializeInc/materialize/issues/17635

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: ci/plugins/mzcompose/hooks/pre-exit [hunk 1]
```diff
@@ -38,6 +62,22 @@ buildkite-agent artifact upload ps-aux.log
...
+# Test 2
+known_issue 17616
+
+if grep "error: missing record 0: Record {" services.log; then
+  known_issue 16778
+fi
+if grep "thread 'coordinator' panicked at 'can't persist timestamp: Error { kind: Stash(StashError { inner: Other(\"unexpected lower\") }) }'" services.log; then
+  known_issue 16927
+fi
+if grep "thread 'coordinator' panicked at 'reference to absent collection: CollectionMissing(User(343))'" services.log; then
```
Comment (philip-stoev) [inline ci/plugins/mzcompose/hooks/pre-exit:77]: Just CollectionMissing would be enough, the number in User() will change across invocations.

**Comment by def-** [comment]
@philip-stoev This is not complete yet, what do you think about this approach?
I added run.log as an artifact too so that we can grep through it as well.
If you think this approach is fine, I'd go ahead and extend it to the remaining CI failures and the cloudtest framework.

**Comment by philip-stoev** [comment]
@def- you are welcome to use this PR to tune your annotation output and such (e.g. run Nightly against it https://www.notion.so/materialize/How-to-do-various-things-8f614c5025ab4defa2a06a1fddc671b1#e7996ef7dae448d385e4e4cabda0b36e) , however I think we need a more general solution that would not require keeping an independent list of known failures. Thankfully, GitHub issues can be used as a the source of truth -- we can annotate each GitHub issue with the error message it presents as. See my proposal here https://github.com/MaterializeInc/materialize/issues/17635#issuecomment-1434544961

File: ci/plugins/mzcompose/hooks/pre-exit [hunk 2]
```diff
@@ -38,6 +60,18 @@ buildkite-agent artifact upload ps-aux.log
 docker ps -a --no-trunc > docker-ps-a.log
 buildkite-agent artifact upload docker-ps-a.log
 
+if grep "error: missing record 0: Record {" services.log; then
+  known_issue 16778
+fi
+if grep "thread 'coordinator' panicked at 'can't persist timestamp: Error { kind: Stash(StashError { inner: Other(\"unexpected lower\") }) }'" services.log; then
```
Comment (philip-stoev) [inline ci/plugins/mzcompose/hooks/pre-exit:66]: Regexps should ideally be as simple as possible, so that they continue to match in case of unrelated changes to the output. So here, `unexpected lower` would have been enough.

File: ci/plugins/mzcompose/hooks/pre-exit [hunk 3]
```diff
@@ -38,6 +60,18 @@ buildkite-agent artifact upload ps-aux.log
...
+  known_issue 16927
+fi
+if grep "thread 'coordinator' panicked at 'reference to absent collection: CollectionMissing(User(" services.log; then
```
Comment (philip-stoev) [inline ci/plugins/mzcompose/hooks/pre-exit:69]: Same here, `CollectionMissing` would have been enough

File: ci/plugins/mzcompose/hooks/pre-exit [hunk 4]
```diff
@@ -38,6 +60,18 @@ buildkite-agent artifact upload ps-aux.log
 docker ps -a --no-trunc > docker-ps-a.log
 buildkite-agent artifact upload docker-ps-a.log
 
+if grep "error: missing record 0: Record {" services.log; then
```
Comment (philip-stoev) [inline ci/plugins/mzcompose/hooks/pre-exit:63]: This one would not be in services.log, but in the `run.log`. Therefore, it is important to fish for interesting lines in all logs.

An example where an error can switch from one log file to another would be your own https://github.com/MaterializeInc/materialize/issues/17616 -- it is now a panic, so is in `services.log`, but if Alexander converts it into a SQL error to be reported to the client, it will end up in the `run.log`

**Comment by def-** [comment]
@philip-stoev Done, I hope something like this makes sense: https://github.com/MaterializeInc/materialize/pull/17688/commits/735682d5845556e9271151d86fd096d025bf8e04

**Comment by philip-stoev** [comment]
@def- thank you, will review as soon as I can.

**Comment by def-** [comment]
> fatal: python v3.9.0+ required

Is there a good way of using Python 3.9 on CI?

File: misc/python/materialize/cli/ci_flake_detect.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import sys
+from typing import Any, List
+
+import junit_xml
+import requests
+
+from materialize import ci_util
+
+CI_RE = re.compile("^ci-regexp: (.*)$", flags=re.MULTILINE)
+FAILURE_RE = re.compile("(panicked|Out of memory)")
```
Comment (philip-stoev) [inline misc/python/materialize/cli/ci_flake_detect.py:23]: Lets make this regular expression match all the examples in the ticket:
- the casing of "Out of memory" is different if the process being killed is inside a memory-restricted container or not :-( .  
- include "internal error" and "FATAL" for the CRDB errors

File: misc/python/materialize/cli/ci_flake_detect.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+import re
+import sys
+from typing import Any, List
+
+import junit_xml
+import requests
+
+from materialize import ci_util
+
+CI_RE = re.compile("^ci-regexp: (.*)$", flags=re.MULTILINE)
```
Comment (philip-stoev) [inline misc/python/materialize/cli/ci_flake_detect.py:22]: This will match everything until the end of the text box, right? I think it is safer to not use multi-line but instead match everything from `ci-regexp:` to the end of the current line.

File: bin/ci-flake-detect [hunk 7]
```diff
@@ -6,9 +8,7 @@
 # As of the Change Date specified in that file, in accordance with
 # the Business Source License, use of this software will be governed
 # by the Apache License, Version 2.0.
+#
+# ci-flake-detect - Detect known flakes based on Github tags
```
Comment (philip-stoev) [inline bin/ci-flake-detect:12]: I do not think "Flake" is the correct term to use here and further down. I think it should be something along the lines of `LoggedError` or another similar term, because in at least 50% of the cases those are actual product bugs and not test issues.

We as a company unfortunately already made a linguistic crime by the widespread use of the word "flake" (even the GitHub tag is named `ci-flake` :-( ). People say "test X is flaky", which pins the blame squarely on the test and not on the code, even in cases where there is an obvious product bug with a panic in the logs. In our messaging as a QA team we should try to counter that tendency by all available means.

File: misc/python/materialize/cli/ci_flake_detect.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            for line in f:
+                match = FAILURE_RE.search(line)
+                if match:
+                    failure_lines.append(line)
+    return sorted(set(failure_lines))
+
+
+def get_flakes_from_github() -> list[Any]:
+    flakes_json = requests.get(
+        f"https://api.github.com/search/issues?q=type:repo:MaterializeInc/materialize%20type:issue%20state:open%20label:ci-flake"
```
Comment (philip-stoev) [inline misc/python/materialize/cli/ci_flake_detect.py:102]: If the `GITHUB_TOKEN` env variable is available, pass it on with the request in order for GitHub to apply a more relaxed rate limiting.

File: misc/python/materialize/cli/ci_flake_detect.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    ).json()
```
Comment (philip-stoev) [inline misc/python/materialize/cli/ci_flake_detect.py:103]: I guess you need to check for an 200 HTTP error code and , for the time being, fail the CI step if one is not received. We can later just log an error once we have seen how this works in practice.

File: misc/python/materialize/cli/ci_flake_detect.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+                )
+                break
+        else:
+            test_case.add_error_info(message="Unknown failure in logs", output=line)
+        junit_suite.test_cases.append(test_case)
+
+    junit_report = ci_util.junit_report_filename(f"{name}_flakes")
+    with junit_report.open("w") as f:
+        junit_xml.to_xml_report_file(f, [junit_suite])
+    ci_util.upload_junit_report(f"{name}_flakes", junit_report)
```
Comment (philip-stoev) [inline misc/python/materialize/cli/ci_flake_detect.py:86]: I do not think your XML file was picked up by the buildkite plugin. For that to happen, I think the name of the junit file needs to conform to the `artifact_paths: test/testdrive/junit_testdrive_*.xml` that is in effect for the testdrive step in ci/test/pipeline-template.yaml ? I guess the `artifact_paths` must be relaxed . Note that this is a general problem, as all other CI steps in all .yaml files will need to be doing the same thing.

File: ci/plugins/mzcompose/hooks/pre-exit [hunk 11]
```diff
@@ -20,23 +20,19 @@ run() {
...
-
 docker ps -a --no-trunc > docker-ps-a.log
-buildkite-agent artifact upload docker-ps-a.log
+
+ARTIFACTS=(run.log services.log journalctl-merge.log netstat-ant.log netstat-panelot.log ps-aux.log docker-ps-a.log)
+for ARTIFACT in "${ARTIFACTS[@]}"; do
+  buildkite-agent artifact upload "$ARTIFACT"
+done
+
+bin/ci-flake-detect --name=mzcompose "${ARTIFACTS[@]}"
```
Comment (philip-stoev) [inline ci/plugins/mzcompose/hooks/pre-exit:35]: I think the `--name` here must be more specific than just `mzcompose`, e.g. it should come from the `BUILDKITE_STEP_KEY` variable?

File: misc/python/materialize/cli/ci_flake_detect.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def get_failure_lines(log_files: List[str]) -> List[str]:
```
Comment (philip-stoev) [inline misc/python/materialize/cli/ci_flake_detect.py:89]: I think this should return a tuple, the name of the log file and the line within the log file. This will make it slightly easier to debug failures.

File: misc/python/materialize/cli/ci_flake_detect.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    if not failure_lines:
+        print("No failures found in logs")
+        return
+
+    flakes = get_flakes_from_github()
+
+    junit_suite = junit_xml.TestSuite("Flakes")
+
+    junit_report = ci_util.junit_report_filename(f"{name}_flakes")
+    for i, line in enumerate(failure_lines):
```
Comment (philip-stoev) [inline misc/python/materialize/cli/ci_flake_detect.py:69]: The line and the file the line came from must be made a part of the information that is being provided about each failure.

File: ci/plugins/mzcompose/hooks/pre-exit (hunk 11 above)
Comment (philip-stoev) [inline ci/plugins/mzcompose/hooks/pre-exit:35]: In order to avoid the

```


fatal: python v3.9.0+ required
--
  | hint: you have v3.7.16
```

error, this needs to be run inside the `ci-builder` container using the  `bin/ci-builder`.  This is what the `run` shell function that is being used elsewhere in the file does.

File: misc/python/materialize/cli/ci_flake_detect.py (hunk 6 above)
Comment (def-) [reply misc/python/materialize/cli/ci_flake_detect.py:22]: No, ^ and $ are beginning and end of line still:
```
>>> CI_RE = re.compile("^ci-regexp: (.*)$", flags=re.MULTILINE)
>>> CI_RE.search("foo\nci-regexp: asdfgh\nbar\n")
<re.Match object; span=(4, 21), match='ci-regexp: asdfgh'>
```
But actually don't need multiline here, so let me remove it.

File: ci/plugins/mzcompose/hooks/pre-exit (hunk 11 above)
Comment (def-) [reply ci/plugins/mzcompose/hooks/pre-exit:35]: By running it inside the ci-builder, do we still have access to the logs directory?

Comment (philip-stoev) [reply ci/plugins/mzcompose/hooks/pre-exit:35]: I believe so, yes. The logs are all placed in the `pwd` of the buildkite process, which I assume is the directory the materialized repo is checked out in. bin/ci-builder is set so that the `pwd` inside the container is that very same directory. The `ci-builder` container is started with:

```
            --volume "$(pwd):$(pwd)"
            --workdir "$(pwd)"
```

**Comment by def-** [comment]
@philip-stoev Is it possible that the BUILDKITE_AGENT_ACCESS_TOKEN does not allow reading artifacts? If so, could we enable them?
I created an access token manually and it works fine with it:
Screenshot 2023-02-22 at 17 22 15
But with the automated runs I'm getting a permission denied error.

**Comment by def-** [comment]
@philip-stoev If you think it's good now, then I'd try to enable this for the rest of the tests and make it ready.
Output looks like this (from the latest run: https://buildkite.com/materialize/tests/builds/50294#018679fa-5645-48f0-952a-64c8e430e182):
Screenshot 2023-02-22 at 17 42 31

**Comment by philip-stoev** [comment]
@def- yes the output looks good for the time being. You can create a new branch and open a new PR if it will be easier for you to restore the tests and the pipeline.

**Comment by def-** [comment]
It seems like the value for Buildkite Test Analytics key is wrong:
```
KeyError: 'BUILDKITE_TEST_ANALYTICS_API_KEY_CHECKS_RESTART_REDPANDA'
```
Should I use BUILDKITE_PLUGINS to get out the composition for that? It seems other places call `ci_util.junit_report_filename()` with a composition name or with "mzcompose" directly:
```
BUILDKITE_PLUGINS="[{\"./ci/plugins/mzcompose\":{\"args\":[\"--scenario=RestartRedpandaDebezium\",\"--seed=01867a25-5635-4dfb-8b21-93578cfc6531\"],\"composition\":\"platform-checks\"}}]"
```
Or alternatively do we want our own test analytics key for logged errors?

File: ci/plugins/cloudtest/hooks/post-command [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+ARTIFACTS=(kubectl-*.log journalctl-merge.log)
+for ARTIFACT in "${ARTIFACTS[@]}"; do
+  buildkite-agent artifact upload "$ARTIFACT"
+done
+
+bin/ci-builder run stable bin/ci-logged-errors-detect "${ARTIFACTS[@]}"
+LOGGED_ERRORS_ARTIFACT="junit_${BUILDKITE_STEP_KEY}_logged_errors_${BUILDKITE_STEP_ID}.xml"
+if [ -f "$LOGGED_ERRORS_ARTIFACT" ]; then
+  buildkite-agent artifact upload "$LOGGED_ERRORS_ARTIFACT"
+fi
```
Comment (benesch) [inline ci/plugins/cloudtest/hooks/post-command:41]: Materialize style is to use lowercase shell variable names, to leave UPPERCASE variable names for distinguishing environment variables.

```suggestion
# shellcheck disable=SC2024
sudo journalctl --merge > journalctl-merge.log

artifacts=(kubectl-*.log journalctl-merge.log)
for artifact in "${artifacts[@]}"; do
  buildkite-agent artifact upload "$artifact"
done

bin/ci-builder run stable bin/ci-logged-errors-detect "${artifacts[@]}"
logged_errors_artifact="junit_${BUILDKITE_STEP_KEY}_logged_errors_${BUILDKITE_STEP_ID}.xml"
if [ -f "$logged_errors_artifact" ]; then
  buildkite-agent artifact upload "$logged_errors_artifact"
fi
```

File: misc/python/materialize/ci_util/__init__.py [hunk 15]
```diff
@@ -49,25 +50,47 @@ def upload_junit_report(suite: str, junit_report: Path) -> None:
...
+
+    if "CI" not in os.environ:
+        return []
+
+    ui.header(f"Getting artifact information from Buildkite")
+    org = os.environ["BUILDKITE_ORGANIZATION_SLUG"]
+    pipeline = os.environ["BUILDKITE_PIPELINE_SLUG"]
+    build = os.environ["BUILDKITE_BUILD_NUMBER"]
+    job = os.environ["BUILDKITE_JOB_ID"]
+    token = os.environ["BUILDKITE_AGENT_ACCESS_TOKEN"]
```
Comment (benesch) [inline misc/python/materialize/ci_util/__init__.py:85]: So the token here is a [session token](https://buildkite.com/docs/agent/v3/tokens#session-tokens). As far as I can tell, the session token is only valid for use with the ["Agents REST API"](https://buildkite.com/docs/apis/agent-api), but you're trying to use the ["Pipelines REST API"](https://buildkite.com/docs/apis/rest-api/access-token) here.

I think you can just use the agent API here instead:

```py
build_id = os.environ["BUILDKITE_BUILD_ID"]
token = os.environ["BUILDKITE_AGENT_ACCESS_TOKEN"]

res = requests.get(
    f"https://agent.buildkite.com/builds/build_id/artifacts/search",
    headers={"Authorization": f"Bearer {token}"},
)
```

There's no documentation for this API, but the Go source code for the `buildkite artifact download` command, which uses this API, is quite readable: https://github.com/buildkite/agent/blob/ad143e4039beb04a9d17d11896d93cf93b2b2532/api/artifacts.go#L135-L154

File: ci/plugins/cloudtest/hooks/post-command [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+ARTIFACTS=(kubectl-*.log journalctl-merge.log)
+for ARTIFACT in "${ARTIFACTS[@]}"; do
+  buildkite-agent artifact upload "$ARTIFACT"
+done
+
+bin/ci-builder run stable bin/ci-logged-errors-detect "${ARTIFACTS[@]}"
+LOGGED_ERRORS_ARTIFACT="junit_${BUILDKITE_STEP_KEY}_logged_errors_${BUILDKITE_STEP_ID}.xml"
+if [ -f "$LOGGED_ERRORS_ARTIFACT" ]; then
+  buildkite-agent artifact upload "$LOGGED_ERRORS_ARTIFACT"
```
Comment (benesch) [inline ci/plugins/cloudtest/hooks/post-command:40]: It seems odd to have the shell scripts handle the upload of the artifact when the Python handles its creation. Especially annoying to have to have the logic for constructing the name of the artifact in two places.

I would either:

  1. Update pipeline.template.yml to upload `junit_*.xml` unconditionally for all jobs, so that this junit gets picked up too when present.
  2. Have the Python script run `subprocess.runv(["buildkite-agent", "artifact", "upload", ...])` instead of doing this in shell.

**Review by benesch — COMMENTED** [review]
This is so cool!

**Comment by benesch** [comment]
> Or alternatively do we want our own test analytics key for logged errors? I only see 4 test suites on https://buildkite.com/organizations/materialize/analytics currently:

Yeah, I think you probably want a separate test analytics project for logged errors? I might be misunderstanding, I don't think we want to mix these "logged error" pseudo tests in with the existing tests.

**Comment by def-** [comment]
@benesch Alright, added one: https://buildkite.com/organizations/materialize/analytics/suites/logged-errors/setup
Can you add the token to CI as `BUILDKITE_ANALYTICS_TOKEN_LOGGED_ERRORS`?

File: ci/plugins/cloudtest/hooks/post-command (hunk 16 above)
Comment (def-) [reply ci/plugins/cloudtest/hooks/post-command:40]: I was doing 1) first, but then wasn't sure if I need to add `junit_*.xml` to each test.

**Comment by def-** [comment]
Still works: https://buildkite.com/materialize/tests/builds/50333#01867d5c-e45a-44f8-bc41-d2ecac1a886d
Another push to try to fix the artifact issue.

**Comment by def-** [comment]
@philip-stoev This is in a state where nightly can run on it I believe.

File: test/testdrive/array.td [hunk 17]
```diff
@@ -10,6 +10,9 @@
 # This test exercises solely to test arrays at the boundary (e.g., by sending
 # them through pgwire).
 
+> SELECT mz_internal.mz_panic('string I want to end up in the log');
```
Comment (philip-stoev) [inline test/testdrive/array.td:13]: This should be removed before you push to `main`

**Comment by def-** [comment]
Worked on nightly: https://buildkite.com/materialize/nightlies/builds/1848

**Review by philip-stoev — APPROVED** [review]
